File: modules/recipely-assistant-kit/android/src/main/java/net/recipely/assistantkit/Envelope.kt
package net.recipely.assistantkit

import android.util.Base64
import java.security.GeneralSecurityException
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * The AES-256-GCM envelope every `/api/v1` request travels inside, for the path
 * where JavaScript does not exist.
 *
 * The Android counterpart of `Envelope.swift`, deliberately the same shape and
 * the same refusals. A shortcut or an AppFunction can answer without the app
 * being launched, so there is no bridge to borrow `aes-envelope.ts` from; the
 * backend accepts no other format, so the format is spoken natively.
 *
 * Parity is pinned by `__fixtures__/aes-gcm-vectors.json` (bytes from OpenSSL),
 * not by reading the three implementations side by side — they agree until the
 * first multibyte character, which is why the fixture carries a Turkish one.
 *
 * Two shapes, on purpose:
 * - the `ByteArray` half is the cipher, and it touches no Android API, so a
 *   plain JVM unit test can drive it. `android.util.Base64` throws the
 *   "not mocked" stub in unit tests, and a cipher that can only be tested on a
 *   device is a cipher that is tested after it ships.
 * - the `String` half is the wire format, and base64 lives only there.
 *   `android.util.Base64` rather than `java.util.Base64` because the latter is
 *   API 26+ and this module supports minSdk 24.
 */
object Envelope {
  /** AES-256 takes a 32-byte key; the "256" is bits. */
  const val KEY_BYTES = 32
  /** GCM's standard nonce length. The backend writes 12 and reads 12. */
  const val IV_BYTES = 12
  /** GCM tag length, appended to the ciphertext rather than carried beside it. */
  const val AUTH_TAG_BYTES = 16

  private const val TRANSFORMATION = "AES/GCM/NoPadding"
  private const val ALGORITHM = "AES"
  private const val TAG_BITS = AUTH_TAG_BYTES * 8
  private const val HEX_RADIX = 16
  private const val HEX_CHARS_PER_BYTE = 2

  sealed class Failure(message: String) : Exception(message) {
    object BadKeyLength : Failure("key must be 32 bytes of hex")
    object BadIvLength : Failure("iv must be 12 bytes")
    object PayloadShorterThanTag : Failure("payload cannot hold an auth tag")
    object NotBase64 : Failure("payload or iv is not base64")
    object AuthenticationFailed : Failure("auth tag did not verify")
  }

  /** What a sealed envelope looks like on the wire. */
  data class Sealed(val payload: String, val iv: String)

  /** Parses the 64-character hex form the app and CI both carry the key in. */
  fun keyFromHex(hex: String): SecretKeySpec {
    if (hex.length != KEY_BYTES * HEX_CHARS_PER_BYTE) throw Failure.BadKeyLength
    val bytes = ByteArray(KEY_BYTES)
    for (i in 0 until KEY_BYTES) {
      val at = i * HEX_CHARS_PER_BYTE
      val byte = hex.substring(at, at + HEX_CHARS_PER_BYTE).toIntOrNull(HEX_RADIX) ?: throw Failure.BadKeyLength
      bytes[i] = byte.toByte()
    }
    return SecretKeySpec(bytes, ALGORITHM)
  }

  /**
   * Seals JSON text for the wire, drawing a fresh nonce per call — reusing one
   * under the same key is the mistake GCM never recovers from. The nonce is
   * generated here rather than left to the provider, because `Cipher.init`
   * without parameters picks its own and only some providers hand it back.
   */
  fun seal(json: String, key: SecretKeySpec): Sealed {
    val iv = ByteArray(IV_BYTES).also { SecureRandom().nextBytes(it) }
    return Sealed(payload = encode(sealBytes(json, key, iv)), iv = encode(iv))
  }

  /** Opens a payload the backend sealed and returns the JSON text inside. */
  fun open(payload: String, iv: String, key: SecretKeySpec): String =
    openBytes(decode(payload), decode(iv), key)

  /**
   * The cipher itself, with the nonce supplied.
   *
   * `internal` because a caller who chooses the nonce can repeat it, and one
   * repeated nonce under a fixed key is enough to lose the key's integrity
   * guarantee. The parity test needs it: a supplied nonce is what makes the
   * sealed bytes comparable to the fixture byte for byte, which is the only way
   * to check the ENCRYPT direction rather than a round trip that would pass
   * against two implementations of the same mistake.
   */
  internal fun sealBytes(json: String, key: SecretKeySpec, iv: ByteArray): ByteArray {
    if (iv.size != IV_BYTES) throw Failure.BadIvLength
    val cipher = Cipher.getInstance(TRANSFORMATION)
    cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(TAG_BITS, iv))
    // doFinal appends the tag, which is the wire format: ciphertext || tag.
    return cipher.doFinal(json.toByteArray(Charsets.UTF_8))
  }

  internal fun openBytes(sealed: ByteArray, iv: ByteArray, key: SecretKeySpec): String {
    // Both lengths are checked before the cipher sees them: a short nonce is
    // silently accepted by some providers, which turns a malformed message into
    // a plausible-looking failure much further downstream.
    if (iv.size != IV_BYTES) throw Failure.BadIvLength
    if (sealed.size <= AUTH_TAG_BYTES) throw Failure.PayloadShorterThanTag

    return try {
      val cipher = Cipher.getInstance(TRANSFORMATION)
      cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(TAG_BITS, iv))
      String(cipher.doFinal(sealed), Charsets.UTF_8)
    } catch (_: GeneralSecurityException) {
      // A tag mismatch and a malformed box mean the same thing to the caller:
      // do not trust these bytes.
      throw Failure.AuthenticationFailed
    }
  }

  private fun encode(bytes: ByteArray): String = Base64.encodeToString(bytes, Base64.NO_WRAP)

  private fun decode(value: String): ByteArray =
    try {
      Base64.decode(value, Base64.DEFAULT)
    } catch (_: IllegalArgumentException) {
      throw Failure.NotBase64
    }
}

File: modules/recipely-assistant-kit/ios/Envelope.swift
import CryptoKit
import Foundation

/// The AES-256-GCM envelope every `/api/v1` request travels inside, implemented
/// for the path where JavaScript does not exist.
///
/// - Note: **Why a second implementation at all.** An intent that answers
///   without opening the app runs with no React Native bridge (see
///   `docs/os-assistants-plan.md` finding D2), so it cannot borrow
///   `aes-envelope.ts`. The backend will not accept anything else, so the
///   format has to be spoken natively.
/// - Note: **Parity is pinned by a fixture, not by reading both files.**
///   `__fixtures__/aes-gcm-vectors.json` holds bytes produced by OpenSSL; this
///   implementation, the JS one and the Kotlin one are each checked against it.
///   Two implementations of a cipher that were only ever reviewed side by side
///   agree right up to the first multibyte character.
/// - Note: **UTF-8 is part of the contract.** The plaintext is JSON text, and
///   the ASCII vector cannot tell a UTF-8 implementation from a UTF-16 one —
///   which is exactly why the fixture carries a Turkish one.
public enum Envelope {
  /// AES-256 takes a 32-byte key; the "256" is bits.
  public static let keyBytes = 32
  /// GCM's standard nonce length. The backend writes 12 and reads 12.
  public static let ivBytes = 12
  /// GCM tag length, appended to the ciphertext rather than carried beside it.
  public static let authTagBytes = 16

  /// Lowercase and uppercase only — not `Character.isHexDigit`, which also admits
  /// full-width and other Unicode digit forms.
  private static let hexAlphabet = Set("0123456789abcdefABCDEF")

  public enum Failure: Error, Equatable {
    case badKeyLength
    case badIvLength
    case payloadShorterThanTag
    case notBase64
    case authenticationFailed
    case notUtf8
  }

  /// Parses the 64-character hex form the app and CI both carry the key in.
  ///
  /// - Note: The alphabet is checked, not only the length. `UInt8("+a", radix: 16)`
  ///   succeeds — Swift's integer initialisers accept a leading sign — so a length
  ///   check alone accepted `"+a"` repeated 32 times as a key, silently producing
  ///   bytes nobody typed. Kotlin's `toIntOrNull(16)` had the same hole pointing the
  ///   other way (`"-1"` became `0xFF`) and the JS half refused both: three parsers
  ///   with three answers, for a value that must be one key everywhere.
  public static func key(fromHex hex: String) throws -> SymmetricKey {
    guard hex.count == keyBytes * 2, hex.allSatisfy({ hexAlphabet.contains($0) }) else {
      throw Failure.badKeyLength
    }
    var bytes = Data(capacity: keyBytes)
    var index = hex.startIndex
    while index < hex.endIndex {
      let next = hex.index(index, offsetBy: 2)
      guard let byte = UInt8(hex[index ..< next], radix: 16) else { throw Failure.badKeyLength }
      bytes.append(byte)
      index = next
    }
    return SymmetricKey(data: bytes)
  }

  /// Seals JSON text, drawing a fresh nonce per call — reusing one under the
  /// same key is the mistake GCM never recovers from.
  public static func seal(json: String, key: SymmetricKey) throws -> (payload: String, iv: String) {
    try seal(json: json, key: key, nonce: AES.GCM.Nonce())
  }

  /// The same seal with the nonce supplied.
  ///
  /// - Note: Not `public`, because a caller who chooses the nonce can repeat it,
  ///   and one repeated nonce under a fixed key is enough to lose the integrity
  ///   guarantee. The parity harness needs it: a supplied nonce is what makes
  ///   the sealed bytes comparable to the fixture byte for byte, which checks
  ///   the ENCRYPT direction rather than a round trip — and a round trip passes
  ///   happily against two implementations of the same mistake.
  static func seal(json: String, key: SymmetricKey, nonce: AES.GCM.Nonce) throws -> (payload: String, iv: String) {
    guard let plaintext = json.data(using: .utf8) else { throw Failure.notUtf8 }
    let box = try AES.GCM.seal(plaintext, using: key, nonce: nonce)
    // `combined` would prepend the nonce; the wire format carries it in its own
    // field, so the payload is ciphertext || tag and nothing else.
    return (
      payload: (box.ciphertext + box.tag).base64EncodedString(),
      iv: Data(box.nonce).base64EncodedString()
    )
  }

  /// Opens a payload the backend sealed and returns the JSON text inside.
  public static func open(payload: String, iv: String, key: SymmetricKey) throws -> String {
    guard let ivData = Data(base64Encoded: iv), let sealed = Data(base64Encoded: payload) else {
      throw Failure.notBase64
    }
    // Both lengths are checked before the cipher sees them: a short nonce is
    // silently padded by some implementations, which turns a malformed message
    // into a plausible-looking decryption failure much further downstream.
    guard ivData.count == ivBytes else { throw Failure.badIvLength }
    guard sealed.count > authTagBytes else { throw Failure.payloadShorterThanTag }

    let tagAt = sealed.count - authTagBytes
    do {
      let box = try AES.GCM.SealedBox(
        nonce: try AES.GCM.Nonce(data: ivData),
        ciphertext: sealed.prefix(tagAt),
        tag: sealed.suffix(authTagBytes)
      )
      let plaintext = try AES.GCM.open(box, using: key)
      guard let json = String(data: plaintext, encoding: .utf8) else { throw Failure.notUtf8 }
      return json
    } catch let failure as Failure {
      throw failure
    } catch {
      // CryptoKit reports a tag mismatch and a malformed box the same way, and
      // the caller can act on neither differently: the answer is "do not trust
      // these bytes".
      throw Failure.authenticationFailed
    }
  }
}
